src/main/kotlin/com/cognifide/gradle/aem/instance/SatisfyTask.kt

```diff
@@ -46,48 +60,53 @@ open class SatisfyTask : SyncTask() {
-    private fun satisfyPackagesOnInstances(packageGroups: List<FileGroup>) {
         for (packageGroup in packageGroups) {
             logger.info("Satisfying group of packages '${packageGroup.name}'.")
 
-            var shouldAwait = false
+            var anyPackageSatisfied = false
 
-            if (config.deployDistributed) {
-                synchronizeInstances({ sync ->
-                    packageGroup.files.onEach {
-                        if (sync.satisfyPackage(it, { sync.distributePackage(it) })) {
-                            shouldAwait = true
+            synchronizeInstances({ sync ->
+                val packageStates = packageGroup.files.fold(mutableMapOf<File, ListResponse.Package?>(), { states, pkg ->
+                    states[pkg] = sync.determineRemotePackage(pkg); states
+                })
+                val anyPackageSatisfiable = packageStates.any {
+                    sync.isSnapshot(it.key) || it.value == null || !it.value!!.installed
+                }
+
+                if (anyPackageSatisfiable) {
+                    packageGroup.initializer(sync)
+                }
+
+                packageStates.forEach { (pkg, state) ->
+                    when {
+                        sync.isSnapshot(pkg) -> {
+                            logger.info("Satisfying package: $pkg (snapshot).")
+                            sync.deployPackage(pkg)
+                            anyPackageSatisfied = true
                         }
-                    }
-                }, Instance.filter(project, config.deployInstanceAuthorName))
-            } else {
-                synchronizeInstances({ sync ->
-                    packageGroup.files.onEach {
-                        if (sync.satisfyPackage(it, { sync.deployPackage(it) })) {
-                            shouldAwait = true
+                        state == null -> {
+                            logger.info("Satisfying package: $pkg (not uploaded).")
+                            sync.deployPackage(pkg)
+                            anyPackageSatisfied = true
+                        }
+                        !state.installed -> {
+                            logger.info("Satisfying package: $pkg (not installed).")
+                            sync.installPackage(state.path)
+                            anyPackageSatisfied = true
+                        }
+                        else -> {
+                            logger.info("Not satisfying package: $pkg (already installed).")
                         }
                     }
-                })
-            }
+                }
+
+                if (anyPackageSatisfiable) {
+                    packageGroup.finalizer(sync)
+                }
+            }, packageGroup.instances)
 
-            if (shouldAwait) {
-                awaitStableInstances()
+            if (anyPackageSatisfied) {
+                packageGroup.completer()
```

Comment:
this is being called when all packages in each group have been satisfied on all instances; called only if it was some change / some package at least installed